### DISCUSSION TITLE
ServiceRequest as parameter for handler

### DISCUSSION THREAD

**Discussion by bestia-dev** [actix/actix-web#2916] [Q&A] [answered]
A request handler async function can take many different parameters (12) that are `impl FromRequest` using the Extractor.
I would simply like to take one parameter of type ServiceRequest and use the extractors later in the code. In my example I will later extract Form only if needed. The Form needs both fields of ServiceRequest:  HttpRequest and PayLoad.
But ServiceRequest is not `impl FromRequest` and I cannot use it as a parameter in request handler.

Did I miss it, is there a way to take ServiceRequest as a parameter in request handler? 
Just like the middleware functions can?

**Comment by robjtede** [comment]
There's no way to use ServiceRequest in this way. Instead you can use:

```rust
async fn handler(req: HttpRequest, pl: web::Payload) -> actix_web::Result<impl Responder> {
  ...
  Form::from_request(&req, &mut pl.into_inner()).await?
  ...
}
```

**Comment by bestia-dev** [comment, marked as answer]
Thank you. 
Is it possible to use only one parameter instead of two ?
Maybe a new struct `RequestAndPayload` with both fields and `impl FromRequest` ? 
Could this work?
```rust
use actix_utils::future::Ready;
use actix_web::error::Error;

pub struct RequestAndPayload{
    req: actix_web::HttpRequest, 
    pl: actix_web::dev::Payload,    
}

impl actix_web::FromRequest for RequestAndPayload {
    type Error = Error;
    type Future = Ready<Result<Self, Error>>;

    #[inline]
    fn from_request(
        req: &actix_web::HttpRequest,
        pl: &mut actix_web::dev::Payload,
    ) -> Self::Future {
        let rap = RequestAndPayload{
            req : req.to_owned(),
            pl : pl.take(),
        };
        
        actix_utils::future::ok(rap)
    }
}
```
